# Pandas Practice - Part 1!

<img src="panda.gif"> 

[gif source](https://25.media.tumblr.com/b268ecc1e1374a7b9386e98ebe623c30/tumblr_mjcpgxpPKL1rjriezo1_400.gif)


`pandas` is a Python package that is widley used for data cleaning and wrangling in Data Science (and beyond!). It's best used on tabular data: think a table with rows and columns. Typically each row is an observation and each column is an attribute of that observation. 

In this example, we will look at data from police stops - where each row is 1 stop, and each column contains so desciption of that stop - like the time the stop look place, the outcome of the stop, etc.

We will explore a few common data cleaning practices using pandas functions like `.drop()`, `.rename()`, `.value_counts()`, and `.astype()` while using pandas documentation to guide us - *remember the using documentation is a critical data science skill* and part of the process!

The dataset here comes from [Kaggle](https://www.kaggle.com/datasets/melihkanbay/police) and is intended to be used to practice using pandas.

In [1]:
# if using Google Collab - uncomment the following lines and run them. 
#
# You'll also need to add the entire path of your data file in your google drive.
#     > You can use the 'files' tab on the left to find your data file, then right click and copy path to get the entire path
#     > Pass the path in "" inside the parameters when you read in the data.


# from google.colab import drive
# drive.mount('/content/drive/',  force_remount=True)

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/ali-rivera/Python-Support-Hours/blob/main/Week3/Week3_BlankResources.ipynb)

In [1]:
import pandas as pd

## Read in data

In [6]:
data = pd.read_csv("https://raw.githubusercontent.com/ali-rivera/Python-Support-Hours/main/5_Pandas1/police.csv")
data.head()

stop_date stop_time  county_name driver_gender  driver_age_raw  \
0  2005-01-02     01:55          NaN             M          1985.0   
1  2005-01-18     08:15          NaN             M          1965.0   
2  2005-01-23     23:15          NaN             M          1972.0   
3  2005-02-20     17:15          NaN             M          1986.0   
4  2005-03-14     10:00          NaN             F          1984.0   

   driver_age driver_race     violation_raw violation  search_conducted  \
0        20.0       White          Speeding  Speeding             False   
1        40.0       White          Speeding  Speeding             False   
2        33.0       White          Speeding  Speeding             False   
3        19.0       White  Call for Service     Other             False   
4        21.0       White          Speeding  Speeding             False   

  search_type   stop_outcome is_arrested stop_duration  drugs_related_stop  
0         NaN       Citation       False      0-15 Min               False  
1         NaN       Citation       False      0-15 Min               False  
2         NaN       Citation       False      0-15 Min               False  
3         NaN  Arrest Driver        True     16-30 Min               False  
4         NaN       Citation       False      0-15 Min               False

In [7]:
data.shape

(91741, 15)

## Drop Redundant Columns

First lets take a look and decide if there are any columns we just don't want. It looks like `driver_age_raw` may be the driver's birth year and age has been added (or calculated) as it's own column. Since this information is redundant, I think we can go ahead and drop `driver_age_raw`.

We can do this using the [.drop()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.drop.html) function.

In [12]:
data.drop(columns="driver_age_raw", inplace=True)
data.head()

stop_date stop_time  county_name driver_gender  driver_age driver_race  \
0  2005-01-02     01:55          NaN             M        20.0       White   
1  2005-01-18     08:15          NaN             M        40.0       White   
2  2005-01-23     23:15          NaN             M        33.0       White   
3  2005-02-20     17:15          NaN             M        19.0       White   
4  2005-03-14     10:00          NaN             F        21.0       White   

      violation_raw violation  search_conducted search_type   stop_outcome  \
0          Speeding  Speeding             False         NaN       Citation   
1          Speeding  Speeding             False         NaN       Citation   
2          Speeding  Speeding             False         NaN       Citation   
3  Call for Service     Other             False         NaN  Arrest Driver   
4          Speeding  Speeding             False         NaN       Citation   

  is_arrested stop_duration  drugs_related_stop  
0       False      0-15 Min               False  
1       False      0-15 Min               False  
2       False      0-15 Min               False  
3        True     16-30 Min               False  
4       False      0-15 Min               False

## Rename for easier indexing

Let's rename some of the the columns to be a bit shorter while still being descriptive. We can use the [.rename()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.rename.html) function to do this.

In [15]:
cols_dict = {'stop_date': 'date',
             'stop_time': 'time',
             'county_name': 'county',
             'driver_gender': 'gender',
             'driver_age': 'age',
             'driver_race': 'race',
             'stop_outcome': 'outcome',
             'is_arrested': 'arrested',
             'drugs_related_stop': 'drug_related'}

data.rename(cols_dict, axis=1, inplace=True)

In [16]:
data.head()

date   time  county gender   age   race     violation_raw violation  \
0  2005-01-02  01:55     NaN      M  20.0  White          Speeding  Speeding   
1  2005-01-18  08:15     NaN      M  40.0  White          Speeding  Speeding   
2  2005-01-23  23:15     NaN      M  33.0  White          Speeding  Speeding   
3  2005-02-20  17:15     NaN      M  19.0  White  Call for Service     Other   
4  2005-03-14  10:00     NaN      F  21.0  White          Speeding  Speeding   

   search_conducted search_type        outcome arrested stop_duration  \
0             False         NaN       Citation    False      0-15 Min   
1             False         NaN       Citation    False      0-15 Min   
2             False         NaN       Citation    False      0-15 Min   
3             False         NaN  Arrest Driver     True     16-30 Min   
4             False         NaN       Citation    False      0-15 Min   

   drug_related  
0         False  
1         False  
2         False  
3         False  
4         False

## Correct dtypes

Before we go any further, let's look at the data type of each column and make sure they are appropriate for the information stored there. There are a few functions we can use to do this, [.info()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.info.html) gives us the most information. [.dtypes](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.dtypes.html) also gives us a list of the data types, if that's all we want.

In [20]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 91741 entries, 0 to 91740
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   date              91741 non-null  object 
 1   time              91741 non-null  object 
 2   county            0 non-null      float64
 3   gender            86406 non-null  object 
 4   age               86120 non-null  float64
 5   race              86408 non-null  object 
 6   violation_raw     86408 non-null  object 
 7   violation         86408 non-null  object 
 8   search_conducted  91741 non-null  bool   
 9   search_type       3196 non-null   object 
 10  outcome           86408 non-null  object 
 11  arrested          86408 non-null  object 
 12  stop_duration     86408 non-null  object 
 13  drug_related      91741 non-null  bool   
dtypes: bool(2), float64(2), object(10)
memory usage: 8.6+ MB


Notice the "RangeIndex" tells use how many rows we have (91741) and "Data Columns" tells us we have 14 columns(after dropping `driver_age_raw`). The "Non-Null Count" column in this outpust shows us that several columns have null values for several rows. We'll keep that in mind for later. For now, let's look a the type of each column.

`stop_date` and `stop_time` are stored as objects, but storing these as a datetime datatype may serve us better. Let's make a new column called `date_time` and store the date AND time in there using the [.to_datetime()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.to_datetime.html#pandas.to_datetime) function.

Now let's take a look at `gender`, `race`, `violation`, `outcome`, `search_type`, and `stop_duration`, which I suspect may be able to be stored as categories...

We can use the [.value_counts()](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.value_counts.html) function for this, which lists out all the values in a column and the count of how many cells hold each value.

In [26]:
cat_cols = ["gender", 'race', 'violation', 'outcome', 'search_type', 'stop_duration']

for col in cat_cols:
    print(data[col].value_counts(), "\n")

gender
M    62895
F    23511
Name: count, dtype: int64 

race
White       62158
Black       12244
Hispanic     9507
Asian        2259
Other         240
Name: count, dtype: int64 

violation
Speeding               48463
Moving violation       16224
Equipment              11020
Other                   4317
Registration/plates     3432
Seat belt               2952
Name: count, dtype: int64 

outcome
Citation            77006
Warning              5294
Arrest Driver        2571
N/D                   590
No Action             589
Arrest Passenger      358
Name: count, dtype: int64 

search_type
Incident to Arrest                                          1219
Probable Cause                                               891
Inventory                                                    220
Reasonable Suspicion                                         197
Protective Frisk                                             161
Incident to Arrest,Inventory                                 129
Incident to Ar

In [27]:
cat_cols.remove('search_type')
cat_cols

['gender', 'race', 'violation', 'outcome', 'stop_duration']

In [30]:
data[cat_cols]data[cat_cols].astype('category')

gender      race         violation        outcome stop_duration
0          M     White          Speeding       Citation      0-15 Min
1          M     White          Speeding       Citation      0-15 Min
2          M     White          Speeding       Citation      0-15 Min
3          M     White             Other  Arrest Driver     16-30 Min
4          F     White          Speeding       Citation      0-15 Min
...      ...       ...               ...            ...           ...
91736      M     White          Speeding        Warning      0-15 Min
91737      F     White         Equipment        Warning      0-15 Min
91738      M     White  Moving violation        Warning      0-15 Min
91739      M     White          Speeding       Citation      0-15 Min
91740      M  Hispanic          Speeding       Citation      0-15 Min

[91741 rows x 5 columns]

It looks like `arrested` is stored as an object but seems to be True/False. Let's take a look at the values and see if we can store it as a bool...

It looks like `arrested` is stored as an object but seems to be True/False. Let's take a look at the values and see if we can store it as a bool...

## Next Steps

Try some of the following (open ended) things on your own:
- Pick a demographic (a certain age range, a race, a gender) and a metric (stop_duration, time of day, outcome, etc). Try to filter on each of these things in 2 different ways.
- `age` is a numeric value. We may want to see the range of ages we have in our dataset. Find a pandas function to get summary statistics on the age variable.
    - Bonus: make a histogram of the age distribution!
- Make age categorical instead of numerical. Pick any number of categories you'd like and the cutoffs. Hint: this is called 'binning' and there is a function to do it! You could also do this with `time`...
    - suggestions: 3 categories - young, middle_aged, and old; by 10s: 10-19, 20-19, 30-39, etc.
- Pick a demographic and *annonomize* it. For example: replace the `gender` category with A/B instead of M/F.
